alton/pause_event.py

```diff
@@ -0,0 +1,366 @@
+"""
+Class which implements the pause/unpause operations for release pipeline systems,
+including backing S3 storage and GoCD integration.
+"""
+from abc import ABCMeta, abstractmethod
+from collections import defaultdict
+from datetime import datetime
+import hashlib
+import logging
+import yaml
+
+import boto
+from boto.exception import S3ResponseError
+from boto.s3.key import Key
+from boto.s3.bucketlistresultset import bucket_lister
+
+from alton.gocd_api import GoCDAPI
+
+
+log = logging.getLogger(__name__)
+
+
+# Official list of pipeline systems.
+# Only these pipeline systems are allowed to be paused or resumed.
+# To add the ability to pause/resume another pipeline system, add it to this dictionary.
+# Keys are pipeline system names.
+# Values are a list of pipeline names to pause/resume when the system is paused/resumed.
+PIPELINE_SYSTEM_INFO = {
+    'edxapp': [
+        'edxapp_release_advancer',
+        'prerelease_edxapp_materials_latest',
+        'edxapp_cut_release_candidate'
+    ]
+}
+
+
+class PauseEventNotFound(Exception):
+    """
+    Raised when a pause event with a specified event ID isn't found.
+    """
+    pass
+
+
+class MultiplePauseEventsFound(Exception):
+    """
+    Raised when multiple pause events are found with the same event ID.
+    """
+    pass
+
+
+class PauseEventOps(object):
+    """
+    Abstract interface class for pausing operations.
+    """
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def add_pipeline_event(self, who_paused, pipeline_system, pause_reason):
```

Comment:
I named them add/remove because they don't necessarily pause/unpause the pipeline system. Only the transition between 0->1 event pauses and the transition between 1->0 events unpauses. I'm open to changing the names - just didn't want to confuse those coming after.